services/api/src/utils/logging.js

```diff
@@ -1,22 +1,133 @@
-  },
-  error(...args) {
-    console.info(...[timeStr(), '-', 'ERROR', '-'].concat(args));
+  base: null,
+  level: config.get('LOG_LEVEL'),
+  timestamp: pino.stdTimeFunctions.isoTime,
+  prettyPrint: process.env.NODE_ENV === 'dev' ? { messageKey: 'message' } : false,
+});
+
+function formatCurrentTrace({ traceId, spanId }) {
+  return {
+    'logging.googleapis.com/spanId': spanId,
+    'logging.googleapis.com/trace': traceId,
+    'logging.googleapis.com/trace_sampled': true,
+  };
+}
+
+const formatters = {
+  gcloud: function ({ request, response, latency }) {
+    // https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry
+    return Object.assign({
+      message: `${request.method} ${request.url} ${response.getHeader('content-length')} - ${latency}ms`,
+      httpRequest: {
+        requestMethod: request.method.toUpperCase(),
+        requestUrl: request.url,
+        requestSize: request.headers['content-length'],
+        status: response.statusCode,
+        userAgent: request.headers['user-agent'],
+        referer: request.headers['referer'],
+        remoteIp: request.headers['x-forwarded-for'],
+        latency: `${Math.floor(latency / 1e3)}.${Math.floor(latency % 1e3)}s`,
+        protocol: request.headers['x-forwarded-proto'],
+        responseSize: response.getHeader && response.getHeader('content-length'),
+      },
+    });
   },
 };
 
-module.exports = {
-  logger,
+exports.loggingMiddleware = function loggingMiddleware({
+  useLevel = 'info',
+  ignoreUserAgents = ['GoogleHC/1.0'],
+  httpRequestFormat = formatters.gcloud,
+}) {
+  function onResFinished(loggerInstance, request, response, error) {
+    const latency = Date.now() - response[startTime];
+    const isError = response.statusCode == 500 || !!error;
+
+    loggerInstance[isError ? 'error' : useLevel](
+      httpRequestFormat({
+        response,
+        request,
+        latency: latency,
+      }),
+      error
+    );
+  }
+
+  function loggingMiddlewareInner(ctx, next) {
+    const res = ctx.res;
+    const req = ctx.req;
+
+    res[startTime] = res[startTime] || Date.now();
+
+    // context only works in prod, as the context is trace based
+    const context = process.env.NODE_ENV === 'production' && getGlobalContext();
+
+    // if the tracer is ignoring certain urls then it wont be available
+    const globalFields = context ? formatCurrentTrace(context) : {};
+
+    const requestLogger = parentLogger.child({
+      ...globalFields,
+    });
+
+    ctx.logger = requestLogger;
+
+    if (ignoreUserAgents.includes(req.headers['user-agent'])) {
+      return next();
+    }
+
+    res.once('finish', () => onResFinished(requestLogger, req, res));
+    res.once('error', (err) => onResFinished(requestLogger, req, res, err));
+
+    return next();
+  }
+  return loggingMiddlewareInner;
+};
+
+/**
+ * @returns {import('pino').Logger} Logger
+ */
+exports.createLogger = function createLogger(options) {
```

Comment:
This may be partly me misunderstanding, but interpreting from the readme and the code, it seems like `logger` is intended to be a singleton, however because you're not always in the context of an HTTP request, the `createLogger` facility is necessary.

I'm wondering if a lot of logic couldn't be removed if it was inverted... ie. `logger` is always a singleton but can handle additional context... a `ctx` option or maybe a `request` and `response` option... something like that. Then the middleware can always pass those options into it via `ctx.logger` as a wrapper.

Again might be misunderstanding but it seems logger as a singleton has access to everything it needs to build up the log and allowing optional context could remove retention of some local vars like `globalContext`, `globalContextFields` etc.